stwcs/updatewcs/astrometry_utils.py

```diff
@@ -0,0 +1,436 @@
+                    hlet_bytes = BytesIO(r_solution.content).getvalue()
+                    hlet = headerlet.Headerlet(file=hlet_bytes)
+                    hlet.init_attrs()
+                    if hlet[0].header['hdrname'] == 'OPUS':
+                        hdrdate = hlet[0].header['date'].split('T')[0]
+                        hlet[0].header['hdrname'] += hdrdate
+                    headerlets[solutionID] = hlet
+            return headerlets, best_solution_id
+
+    def addObservation(self, observationID, new_solution):
+        """Add WCS from current observation to database"""
+        if not self.perform_step:
+            return
+
+        serviceEndPoint = self.serviceLocation+'observation/create'
+        headers = {'Content-Type': 'application/octet-stream'}
+
+        r = requests.post(serviceEndPoint, data=new_solution, headers=headers)
+        if r.status_code == requests.codes.ok:
+            logger.info("AstrometryDB service updated with new entry for {}".
+                        format(observationID))
+        else:
+            l = "Problem encountered when adding {} to database".\
+                           format(observationID)
+            if self.raise_errors:
+                raise Exception(l)
+            else:
+                logger.warning(l)
+
+    def isAvailable(self):
+        """Test availability of astrometryDB web-service."""
+        if not self.perform_step:
+            return
+
+        serviceEndPoint = self.serviceLocation+'availability'
+
+        try:
+            r = requests.get(serviceEndPoint, headers=self.headers)
+
+            if r.status_code == requests.codes.ok:
+                logger.info('AstrometryDB service available...')
+                self.available_code['code'] = r.status_code
+                self.available_code['text'] = 'Available'
+                self.available = True
+            else:
+                logger.warning('WARNING : AstrometryDB service unavailable!')
+                logger.warning('          AstrometryDB called: {}'.format(
+                                self.serviceLocation))
+                logger.warning('          AstrometryDB status: {}'.format(
+                               r.status_code))
+                logger.warning('          AstrometryDB text: {}'.format(
+                               r.text))
+                self.available_code['code'] = r.status_code
+                self.available_code['text'] = r.text
+                self.available = False
+                if self.raise_errors:
+                    e = "AstrometryDB service unavailable!"
+                    raise ConnectionRefusedError(e)
+
+        except Exception as err:
+            logger.warning('WARNING : AstrometryDB service inaccessible!')
+            logger.warning('    AstrometryDB called: {}'.format(
+                                self.serviceLocation))
+            logger.warning('    Status: {}'.format(r.status))
```

Comment:
Since `r` is defined inside the `try` statement when an exception is raised it is not available, i.e. `r` is undefined at this point.